saw-core/src/SAWCore/Term/CtxTerm.hs

```diff
@@ -115,79 +116,68 @@ instance UsesDataType [(LocalName, Term)] where
 -- where the @pi@ are the distinct bound variables bound in the @params@
 -- context, given as argument, and that the @xj@ have no occurrences of @d@. If
 -- the given type is of this form, return the @xj@.
-asCtorDTApp :: Name -> [param] ->
+asCtorDTApp :: Name -> [ExtCns Term] ->
                [index] ->
-               [a] ->
-               [b] ->
                Term ->
                Maybe [Term]
-asCtorDTApp d params dt_ixs ctx1 ctx2 (ctxAsDataTypeApp d params dt_ixs ->
+asCtorDTApp d params dt_ixs (ctxAsDataTypeApp d params dt_ixs ->
                                        Just (param_vars, ixs))
-  | isVarList params ctx1 ctx2 param_vars &&
-    not (any (usesDataType d) ixs)
+  | isVarList params param_vars && not (any (usesDataType d) ixs)
   = Just ixs
   where
-    -- Check that the given list of terms is a list of bound variables, one for
-    -- each parameter, in the context extended by the given arguments
-    isVarList :: [param] ->
-                 [a] ->
-                 [b] ->
-                 [Term] ->
-                 Bool
-    isVarList _ _ _ [] = True
-    isVarList (_ : ps) c1 c2 ((unwrapTermF -> LocalVar i) : ts) =
-      i == length ps + length c1 + length c2 &&
-      isVarList ps c1 c2 ts
-    isVarList _ _ _ _ = False
-asCtorDTApp _ _ _ _ _ _ = Nothing
+    -- Check that the given list of terms is a list of named
+    -- variables, one for each parameter
+    isVarList :: [ExtCns Term] -> [Term] -> Bool
+    isVarList _ [] = True
+    isVarList (p : ps) ((asExtCns -> Just ec) : ts) =
+      ec == p && isVarList ps ts
+    isVarList _ _ = False
+    -- isVarList ps (t : _) = error $ unwords ["isVarList", showTerm t, show (map (showTerm . Unshared . FTermF . ExtCns) ps)]
+asCtorDTApp _ _ _ _ = Nothing
 
 
 -- | Check that an argument for a constructor has one of the allowed forms
-asCtorArg :: Name -> [param] ->
+asCtorArg :: Name -> [ExtCns Term] ->
              [index] ->
-             [prev] ->
              Term ->
              Maybe CtorArg
-asCtorArg d params dt_ixs prevs (asPiList ->
+asCtorArg d params dt_ixs (asPiList ->
                                  (zs,
-                                  asCtorDTApp d params dt_ixs prevs zs ->
+                                  asCtorDTApp d params dt_ixs ->
                                   Just ixs))
```

Comment:
There is a lot of rightward drift in this code. Consider using pattern guards to clean this up a bit, e.g.,

```hs
asCtorArg d params dt_ixs tp
  | (zs, asCtorDTApp d params dt_ixs -> Just ixs)) <- asPiList tp
```

Reply:
Again, this code will be imminently replaced in #2466.